### PR TITLE
Fix gasnet smp support

### PR DESCRIPTION
**Pull request by ronawho**
Fix smp support for gasnet. SMP is a "fake" conduit that uses shared-memory to
imitate a network. This allows users to try out gasnet and multi-locale
programs on a single node. It's similar to using amudp with local spawning, but
it doesn't require sockets and is simpler to setup.

The main change was ensuring that we don't disable pshm support for smp since
it's required. We also don't want to set a segment since there is no segment
for smp. Lastly we needed to add a simple launcher which effectively just sets
GASNET_PSHM_NODES. It would have been cleaner to do this in the runtime, but
unfortunately we don't parse `--numLocales/-nl` until after comm init is done.

This fixes https://github.com/chapel-lang/chapel/issues/7569

### CODE REVIEW

**Comment by ronawho** [comment]
Passes a full paratest. I'll add some nightly testing in the next day or two to make sure this continues to work.

**Comment by ronawho** [comment]
@gbtitus and maybe @bradcray mind reviewing? Probably easiest to review commit at a time.

Any documentation I'm missing? It didn't really feel worthwhile to add a doc/rst/platforms/smp.rst, but I could be talked into that.

File: runtime/src/launch/smp/launch-smp.c [hunk 1]
```diff
@@ -20,53 +20,24 @@
...
-  if (strlen(command)+1 > size) {
-    chpl_internal_error("buffer overflow");
-  }
-
-  return command;
-}
-
-
 int chpl_launch(int argc, char* argv[], int32_t numLocales) {
+  char baseCommand[4096];
```
[the remark this answers is not in the document]
Comment (ronawho) [reply runtime/src/launch/smp/launch-smp.c]: I stole this from another launcher, they're all pretty terrible. I don't think it's worth spending much time improving our existing launchers and would much rather see any significant effort put towards replacing our launchers with a scripting infrastructure. 90% of the launcher code is string manipulation and launching subprocesses, both of which are ugly in C.

File: runtime/src/launch/smp/launch-smp.c [hunk 2]
```diff
@@ -20,53 +20,24 @@
...
   char numlocalesval[11]; // big enough for int32_t
   snprintf(numlocalesval, sizeof(numlocalesval), "%d", (int)numLocales);
   if (setenv("GASNET_PSHM_NODES", numlocalesval, 1) != 0) {
     chpl_error("Cannot setenv(\"GASNET_PSHM_NODES\")", 0, 0);
   }
 
-  return chpl_launch_using_system(chpl_launch_create_command(argc, argv, numLocales),
-                                  argv[0]);
+  chpl_compute_real_binary_name(argv[0]);
+  sprintf(baseCommand, "%s", chpl_get_real_binary_name());
```
[the remark this answers is not in the document]
Comment (ronawho) [reply runtime/src/launch/smp/launch-smp.c]: I'll change it, but this was also copied from another launcher. Not to say that makes it ok, but it's in equally terrible company

**Comment by ronawho** [comment]
>You could avoid the launcher by adding another arg parsing mode (see runtime/include/arg.h:parse_dash_E for example) that would give early access to the -nl arg value. That would be less total code than what is here, but of course what is here is already done! There is one line-comment that needs addressing unless you do dump the launcher.

I tried this and I wasn't able to do it in ~2 hours. The launcher code, while not very satisfying, was really simple to do since it's mostly just copying boilerplate code. That said, I'm not very familiar with the arg parsing code, so maybe it wouldn't take so long if I was familiar with it :)

**Comment by bradcray** [comment]
I feel mildly allergic to adding specialized "early argument parsing" passes because (as a somewhat artificial example) it seems too easy to mis-parse `--compopts -nl=2` as "Oh they said to use 2 locales" rather than "Oh, they said that `-nl=2` should be stored in the variable `compopts` given:

```chapel
config var compopts: string;
```

File: runtime/src/launch/smp/launch-smp.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (bradcray) [reply runtime/src/launch/smp/launch-smp.c]: I tend to agree with the "invest in script-based launchers over fixing the existing launchers" philosophy, at least as long as they continue to work "well enough for now."

[the remark this answers is not in the document]
Comment (gbtitus) [reply runtime/src/launch/smp/launch-smp.c]: I'm totally with you, but where I see the possibility of the most user-hostile failure possible, namely a segfault, and I think I can easily turn that into, say, an internal error, I'm inclined to fix it.  Fortunately (for me) in this case I've convinced myself that the binary path handling in the main launcher won't actually segfault.

**Pull request merged**